File: src/GoogologyLib/Notations/Number/Conway.cs
using GoogologyLib.Core;

namespace GoogologyLib.Notations.Number;

/// <summary>Conway chained-arrow notation (康威链式箭头).</summary>
public class Conway : Notation
{
    private List<CNode> _chain = [];

    public Conway() { }
    public Conway(string s) => StringToIt(s);

    public override string Name => "conway";
    public override Family Family => Core.Family.Number;
    public override string Creator => "John Conway";
    public override string Version => "1";

    public override Capabilities Capabilities
    {
        get
        {
            var c = new Capabilities();
            c.Set(Op.FromString);
            c.Set(Op.ToString);
            c.Set(Op.Expand);
            c.Set(Op.ExpandTo);
            return c;
        }
    }

    public override void StringToIt(string s)
    {
        var t = s.Replace("→", "->").Replace(" ", "");
        _chain = t.Split("->")
            .Select(part => CNode.Value(long.Parse(part)))
            .ToList();
    }

    public override string ToLatex() => Ser(_chain);

    public override Notation Expand(int n = 1)
    {
        for (int i = 0; i < n; i++)
            _chain = StepOnce(_chain);
        return this;
    }

    public override Notation ExpandTo(int length)
    {
        var prev = ToLatex();
        while (true)
        {
            Expand(1);
            var cur = ToLatex();
            if (cur.Length >= length || cur == prev) return this;
            prev = cur;
        }
    }

    // --- internal ---

    private class CNode
    {
        public bool IsInt;
        public long Val;
        public List<CNode>? Sub;

        public static CNode Value(long v) => new() { IsInt = true, Val = v };
        public static CNode SubChain(List<CNode> s) => new() { IsInt = false, Sub = s };
    }

    private static string Ser(List<CNode> ch)
    {
        if (ch.Count == 2 && ch[0].IsInt && ch[1].IsInt)
            return $"{ch[0].Val}^{{{ch[1].Val}}}";
        var parts = ch.Select(node =>
            node.IsInt ? node.Val.ToString() : $"({Ser(node.Sub!)})");
        return string.Join(" \\rightarrow ", parts);
    }

    private static List<CNode> StepOnce(List<CNode> ch)
    {
        var n = ch.Count;
        if (n <= 2)
        {
            if (n <= 1) return ch;
            if (ch[0].IsInt && ch[1].IsInt) return ch;  // terminal: a -> b
            return RecurseSub(ch);
        }

        // rule 2a: trailing 1  (X -> a -> 1 = X -> a)
        if (ch[n - 1].IsInt && ch[n - 1].Val == 1)
            return ch.Take(n - 1).ToList();

        // rule 2b: middle 1  (X -> 1 -> Y = X)
        if (ch[n - 2].IsInt && ch[n - 2].Val == 1)
            return ch.Take(n - 2).ToList();

        // rule 3: X -> a -> b = X -> (X -> a-1 -> b) -> b-1
        if (ch[n - 1].IsInt && ch[n - 2].IsInt)
        {
            var a = ch[n - 2].Val;
            var b = ch[n - 1].Val;
            var x = ch.Take(n - 2).ToList();
            var inner = x.Concat([CNode.Value(a - 1), CNode.Value(b)]).ToList();
            return x.Concat([CNode.SubChain(inner), CNode.Value(b - 1)]).ToList();
        }

        return RecurseSub(ch);
    }

    private static List<CNode> RecurseSub(List<CNode> ch)
    {
        var changed = false;
        var r = ch.Select(node =>
        {
            if (!node.IsInt)
            {
                var before = Ser(node.Sub!);
                var s = StepOnce(node.Sub!);
                var after = Ser(s);
                if (after != before)
                {
                    changed = true;
                    if (s.Count == 1 && s[0].IsInt)
                        return CNode.Value(s[0].Val);
                    return CNode.SubChain(s);
                }
            }
            return node;
        }).ToList();
        return changed ? r : ch;
    }
}

File: src/main/java/googology/notations/number/Conway.java
package googology.notations.number;

import java.util.*;
import java.util.stream.*;
import googology.core.*;

/** Conway chained-arrow notation (康威链式箭头). */
public class Conway extends Notation {
    private List<CNode> chain = new ArrayList<>();

    public Conway() {}
    public Conway(String s) { stringToIt(s); }

    @Override public String name() { return "conway"; }
    @Override public Family family() { return Family.NUMBER; }
    @Override public String creator() { return "John Conway"; }
    @Override public String version() { return "1"; }

    @Override
    public Capabilities capabilities() {
        var c = new Capabilities();
        c.set(Op.FROM_STRING);
        c.set(Op.TO_STRING);
        c.set(Op.EXPAND);
        c.set(Op.EXPAND_TO);
        return c;
    }

    @Override
    public void stringToIt(String s) {
        var t = s.replace("→", "->").replace(" ", "");
        chain = Arrays.stream(t.split("->"))
            .map(part -> CNode.value(Long.parseLong(part)))
            .collect(Collectors.toList());
    }

    @Override
    public String toLatex() { return ser(chain); }

    @Override
    public Notation expand(int n) {
        for (int i = 0; i < n; i++) chain = stepOnce(chain);
        return this;
    }

    @Override
    public Notation expandTo(int length) {
        String prev = toLatex();
        while (true) {
            expand(1);
            String cur = toLatex();
            if (cur.length() >= length || cur.equals(prev)) return this;
            prev = cur;
        }
    }

    // --- internal ---

    private static class CNode {
        boolean isInt;
        long val;
        List<CNode> sub;

        static CNode value(long v) { var n = new CNode(); n.isInt = true; n.val = v; return n; }
        static CNode subChain(List<CNode> s) { var n = new CNode(); n.isInt = false; n.sub = s; return n; }
    }

    private static String ser(List<CNode> ch) {
        if (ch.size() == 2 && ch.get(0).isInt && ch.get(1).isInt)
            return ch.get(0).val + "^{" + ch.get(1).val + "}";
        return ch.stream()
            .map(node -> node.isInt ? Long.toString(node.val) : "(" + ser(node.sub) + ")")
            .collect(Collectors.joining(" \\rightarrow "));
    }

    private static List<CNode> stepOnce(List<CNode> ch) {
        int n = ch.size();
        if (n <= 2) {
            if (n <= 1) return ch;
            if (ch.get(0).isInt && ch.get(1).isInt) return ch;  // terminal
            return recurseSub(ch);
        }

        // rule 2a: trailing 1
        if (ch.get(n - 1).isInt && ch.get(n - 1).val == 1)
            return ch.subList(0, n - 1);

        // rule 2b: middle 1
        if (ch.get(n - 2).isInt && ch.get(n - 2).val == 1)
            return ch.subList(0, n - 2);

        // rule 3: X -> a -> b = X -> (X -> a-1 -> b) -> b-1
        if (ch.get(n - 1).isInt && ch.get(n - 2).isInt) {
            long a = ch.get(n - 2).val;
            long b = ch.get(n - 1).val;
            var x = ch.subList(0, n - 2);
            var inner = new ArrayList<>(x);
            inner.add(CNode.value(a - 1));
            inner.add(CNode.value(b));
            var result = new ArrayList<>(x);
            result.add(CNode.subChain(inner));
            result.add(CNode.value(b - 1));
            return result;
        }

        return recurseSub(ch);
    }

    private static List<CNode> recurseSub(List<CNode> ch) {
        boolean changed = false;
        var r = new ArrayList<CNode>();
        for (var node : ch) {
            if (!node.isInt) {
                String before = ser(node.sub);
                var s = stepOnce(node.sub);
                String after = ser(s);
                if (!after.equals(before)) {
                    changed = true;
                    if (s.size() == 1 && s.get(0).isInt)
                        r.add(CNode.value(s.get(0).val));
                    else
                        r.add(CNode.subChain(s));
                    continue;
                }
            }
            r.add(node);
        }
        return changed ? r : ch;
    }
}